File: Problem47_PathSumII.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> pathsList = new ArrayList<List<Integer>>();
        List<Integer> pathNodes = new ArrayList<>();
        this.recurseTree(root, targetSum, pathNodes, pathsList);
        return pathsList;
    }
    private void recurseTree(TreeNode node, int remainingSum, List<Integer> pathNodes, List<List<Integer>> pathsList) {
        if(node == null)
            return;
        
        pathNodes.add(node.val);
        if(remainingSum == node.val && node.left == null && node.right == null)
            pathsList.add(new ArrayList<>(pathNodes));
        else {
            this.recurseTree(node.left, remainingSum - node.val, pathNodes, pathsList);
            this.recurseTree(node.right, remainingSum - node.val, pathNodes, pathsList);
        }
        pathNodes.remove(pathNodes.size() - 1);
    }
}

File: Problem-47 Path Sum II.py
# 113. Path Sum II
# https://leetcode.com/problems/path-sum-ii/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Iterative Approach
# Time complexiety: O(n)
# Space complexiety: O(n)
class Solution:    
    def pathSum(self, root, targetSum):
        if not root:
            return []
        
        res = []
        
        stack = [(root,[root.val], targetSum-root.val)]
        
        while stack:
            root, eleList, t = stack.pop()
            
            if root.left == None and root.right == None and t == 0:
                res.append(eleList)
            
            if root.right:
                stack.append((root.right,eleList+[root.right.val],t-root.right.val))
            
            if root.left:
                stack.append((root.left,eleList+[root.left.val],t-root.left.val))
        return res
